File: kernel/mm/pmm.c
#include "pmm.h"
#include "klog.h"
#include <stddef.h>

/* Physical memory manager: a one-bit-per-page bitmap over all of
 * physical memory, tracking which 4KiB pages are in use. All pages
 * start "used"; the memory map's usable entries flip their pages to
 * "free". The bitmap itself lives in the first usable region (mapped
 * through the HHDM), and its own pages are kept marked as used.
 *
 * Allocation is first-fit with a moving hint, which is simple and
 * good enough until a more sophisticated allocator is needed.
 *
 * Note: Limine has already set up paging (kernel in the higher half,
 * all physical memory mapped at HHDM offset), so no page tables are
 * touched here — this manager tracks *ownership* of physical frames
 * only. */

/* bit set = page in use */
static uint64_t *pmm_bitmap = NULL;
static uint64_t pmm_total_pages = 0;
static uint64_t pmm_first_free_hint = 0;
static uint64_t pmm_free_count = 0;

static inline int bitmap_test(uint64_t page) {
    return (pmm_bitmap[page / 64] >> (page % 64)) & 1u;
}

static inline void bitmap_set(uint64_t page) {
    pmm_bitmap[page / 64] |= (1ull << (page % 64));
}

static inline void bitmap_clear(uint64_t page) {
    pmm_bitmap[page / 64] &= ~(1ull << (page % 64));
}

void pmm_init(const struct limine_memmap_response *memmap, uint64_t hhdm) {
    if (memmap == NULL || memmap->entries == NULL) {
        klog("PMM: FATAL no memory map\n");
        for (;;) { __asm__ volatile ("cli; hlt"); }
    }

    /* Find the highest address described by the map — we track
     * physical pages up to that point. Only *usable* entries count:
     * non-usable entries (BIOS holes, bootloader data, ACPI tables,
     * and the huge reserved ranges some firmware reports at the top
     * of the address space) are never allocated from, and Limine's
     * HHDM mapping does not cover every one of their pages (e.g. the
     * 0xA0000 VGA hole), so walking them would page-fault. */
    uint64_t highest = 0;
    for (uint64_t i = 0; i < memmap->entry_count; i++) {
        struct limine_memmap_entry *e = memmap->entries[i];
        if (e->type != LIMINE_MEMMAP_USABLE) continue;
        uint64_t end = e->base + e->length;
        if (end > highest) highest = end;
    }

    pmm_total_pages = (highest + PAGE_SIZE - 1) / PAGE_SIZE;

    uint64_t bitmap_bytes = pmm_total_pages / 8;
    uint64_t bitmap_pages = (bitmap_bytes + PAGE_SIZE - 1) / PAGE_SIZE;

    /* Place the bitmap at the start of the first usable region. A
     * usable region legitimately starting at physical address 0 (which
     * UEFI boot can report, under protocol base revision >= 3) must
     * not be mistaken for "no usable region", so track success with a
     * flag rather than testing the address itself. */
    uintptr_t bitmap_phys = 0;
    int bitmap_found = 0;
    for (uint64_t i = 0; i < memmap->entry_count; i++) {
        struct limine_memmap_entry *e = memmap->entries[i];
        if (e->type == LIMINE_MEMMAP_USABLE) {
            bitmap_phys = e->base;
            bitmap_found = 1;
            break;
        }
    }
    if (!bitmap_found) {
        klog("PMM: FATAL no usable memory region\n");
        for (;;) { __asm__ volatile ("cli; hlt"); }
    }

    pmm_bitmap = (uint64_t *)(hhdm + bitmap_phys);

    /* Start with every tracked page used. */
    for (uint64_t i = 0; i < (bitmap_bytes + 7) / 8; i++) {
        pmm_bitmap[i] = ~0ull;
    }

    /* Free the pages described as usable by the bootloader. */
    pmm_free_count = 0;
    for (uint64_t i = 0; i < memmap->entry_count; i++) {
        struct limine_memmap_entry *e = memmap->entries[i];
        if (e->type != LIMINE_MEMMAP_USABLE) continue;

        uint64_t start_page = e->base / PAGE_SIZE;
        uint64_t end_page = (e->base + e->length + PAGE_SIZE - 1) / PAGE_SIZE;
        if (end_page > pmm_total_pages) end_page = pmm_total_pages;

        for (uint64_t p = start_page; p < end_page; p++) {
            if (!bitmap_test(p)) continue;   /* not double-freeing */
            bitmap_clear(p);
            pmm_free_count++;
        }
    }

    /* Reserve the pages the bitmap itself occupies (first usable
     * region) plus anything below 1MiB — the bootloader maps / BIOS
     * data and real-mode bounce space live down there. */
    uint64_t bitmap_start_page = bitmap_phys / PAGE_SIZE;

    uint64_t low_reserve_end = 0x100000 / PAGE_SIZE;   /* 1 MiB */
    for (uint64_t p = 0; p < low_reserve_end; p++) {
        if (!bitmap_test(p)) { bitmap_set(p); pmm_free_count--; }
    }
    for (uint64_t p = bitmap_start_page; p < bitmap_start_page + bitmap_pages; p++) {
        if (!bitmap_test(p)) { bitmap_set(p); pmm_free_count--; }
    }

    pmm_first_free_hint = bitmap_start_page + bitmap_pages;

    klog("PMM: %lu pages total, %lu free, bitmap %lu pages @ phys %lx\n",
         (unsigned long)pmm_total_pages, (unsigned long)pmm_free_count,
         (unsigned long)bitmap_pages, (unsigned long)bitmap_phys);
}

uintptr_t pmm_alloc_page(void) {
    if (pmm_bitmap == NULL) return 0;

    /* Scan from the hint forward, then wrap. */
    for (uint64_t pass = 0; pass < 2; pass++) {
        uint64_t start = pass == 0 ? pmm_first_free_hint : 0;
        uint64_t end   = pass == 0 ? pmm_total_pages : pmm_first_free_hint;
        for (uint64_t p = start; p < end; p++) {
            if (!bitmap_test(p)) {
                bitmap_set(p);
                pmm_free_count--;
                pmm_first_free_hint = p + 1;
                return p * PAGE_SIZE;
            }
        }
    }
    klog("PMM: out of physical memory\n");
    return 0;
}

void pmm_free_page(uintptr_t paddr) {
    if (paddr == 0 || (paddr % PAGE_SIZE) != 0) return;
    uint64_t p = paddr / PAGE_SIZE;
    if (p >= pmm_total_pages || !bitmap_test(p)) return;  /* out of range or already free */
    bitmap_clear(p);
    pmm_free_count++;
    if (p < pmm_first_free_hint) pmm_first_free_hint = p;
}

uint64_t pmm_free_page_count(void) {
    return pmm_free_count;
}

File: rust/src/memory/pmm.rs
//! Physical memory manager: a one-bit-per-page bitmap over all of
//! physical memory, tracking which 4 KiB pages are in use.
//!
//! Faithful port of `kernel/mm/pmm.c`: all pages start "used"; the
//! memory map's usable entries flip their pages to "free"; the bitmap
//! itself lives in the first usable region (mapped through the HHDM)
//! and its own pages stay marked used. Allocation is first-fit with a
//! moving hint.
//!
//! Note: Limine has already set up paging (kernel in the higher half,
//! all physical memory mapped at the HHDM offset), so no page tables
//! are touched here — this manager tracks *ownership* of physical
//! frames only.

use crate::boot::requests::{LimineMemmapResponse, LIMINE_MEMMAP_USABLE};
use crate::memory::hhdm::Hhdm;
use crate::memory::PAGE_SIZE;

/// bit set = page in use
static mut PMM_BITMAP: *mut u64 = core::ptr::null_mut();
static mut PMM_TOTAL_PAGES: u64 = 0;
static mut PMM_FIRST_FREE_HINT: u64 = 0;
static mut PMM_FREE_COUNT: u64 = 0;

#[inline]
unsafe fn bitmap_test(page: u64) -> bool {
    let word = unsafe { *PMM_BITMAP.add((page / 64) as usize) };
    (word >> (page % 64)) & 1 != 0
}

#[inline]
unsafe fn bitmap_set(page: u64) {
    let word = unsafe { &mut *PMM_BITMAP.add((page / 64) as usize) };
    *word |= 1u64 << (page % 64);
}

#[inline]
unsafe fn bitmap_clear(page: u64) {
    let word = unsafe { &mut *PMM_BITMAP.add((page / 64) as usize) };
    *word &= !(1u64 << (page % 64));
}

/// Builds the physical page allocator from the bootloader's memory
/// map. `hhdm` is the higher-half direct-map handle. Must be called
/// once, after the Limine requests are populated.
pub fn init(hhdm: &Hhdm, memmap: &LimineMemmapResponse) {
    unsafe {
        if memmap.entries.is_null() {
            crate::kprintln!("[mm] PMM: FATAL no memory map");
            crate::kernel::halt();
        }

        /* Find the highest address described by the map — we track
         * physical pages up to that point. Only *usable* entries count:
         * non-usable entries (BIOS holes, bootloader data, ACPI tables,
         * and the huge reserved ranges some firmware reports at the top
         * of the address space) are never allocated from, and Limine's
         * HHDM mapping does not cover every one of their pages (e.g. the
         * 0xA0000 VGA hole), so walking them would page-fault. */
        let mut highest = 0u64;
        for i in 0..memmap.entry_count as usize {
            let e = &**memmap.entries.add(i);
            if e.type_ != LIMINE_MEMMAP_USABLE {
                continue;
            }
            let end = e.base + e.length;
            if end > highest {
                highest = end;
            }
        }

        PMM_TOTAL_PAGES = (highest + PAGE_SIZE - 1) / PAGE_SIZE;

        let bitmap_bytes = PMM_TOTAL_PAGES / 8;
        let bitmap_pages = (bitmap_bytes + PAGE_SIZE - 1) / PAGE_SIZE;

        /* Place the bitmap at the start of the first usable region. A
         * usable region legitimately starting at physical address 0
         * must not be mistaken for "no usable region", so track success
         * with a flag rather than testing the address itself. */
        let mut bitmap_phys = 0u64;
        let mut bitmap_found = false;
        for i in 0..memmap.entry_count as usize {
            let e = &**memmap.entries.add(i);
            if e.type_ == LIMINE_MEMMAP_USABLE {
                bitmap_phys = e.base;
                bitmap_found = true;
                break;
            }
        }
        if !bitmap_found {
            crate::kprintln!("[mm] PMM: FATAL no usable memory region");
            crate::kernel::halt();
        }

        PMM_BITMAP = hhdm.phys_to_virt(bitmap_phys) as *mut u64;

        /* Start with every tracked page used. */
        let words = (bitmap_bytes + 7) / 8;
        for i in 0..words as usize {
            *PMM_BITMAP.add(i) = !0u64;
        }

        /* Free the pages described as usable by the bootloader. */
        PMM_FREE_COUNT = 0;
        for i in 0..memmap.entry_count as usize {
            let e = &**memmap.entries.add(i);
            if e.type_ != LIMINE_MEMMAP_USABLE {
                continue;
            }
            let start_page = e.base / PAGE_SIZE;
            let mut end_page = (e.base + e.length + PAGE_SIZE - 1) / PAGE_SIZE;
            if end_page > PMM_TOTAL_PAGES {
                end_page = PMM_TOTAL_PAGES;
            }
            for p in start_page..end_page {
                if bitmap_test(p) {
                    /* not double-freeing */
                    bitmap_clear(p);
                    PMM_FREE_COUNT += 1;
                }
            }
        }

        /* Reserve the pages the bitmap itself occupies (first usable
         * region) plus anything below 1 MiB — the bootloader maps / BIOS
         * data and real-mode bounce space live down there. */
        let bitmap_start_page = bitmap_phys / PAGE_SIZE;

        let low_reserve_end = 0x100000 / PAGE_SIZE; /* 1 MiB */
        for p in 0..low_reserve_end {
            if !bitmap_test(p) {
                bitmap_set(p);
                PMM_FREE_COUNT -= 1;
            }
        }
        for p in bitmap_start_page..bitmap_start_page + bitmap_pages {
            if !bitmap_test(p) {
                bitmap_set(p);
                PMM_FREE_COUNT -= 1;
            }
        }

        PMM_FIRST_FREE_HINT = bitmap_start_page + bitmap_pages;

        let total_pages = PMM_TOTAL_PAGES;
        let free_count = PMM_FREE_COUNT;
        crate::kprintln!(
            "[mm] PMM: {} pages total, {} free, bitmap {} pages @ phys {:#x}",
            total_pages,
            free_count,
            bitmap_pages,
            bitmap_phys
        );
    }
}

/// Returns the physical address of one free 4 KiB page, or 0 if out of
/// memory. The returned frame is marked allocated.
pub fn alloc_page() -> u64 {
    unsafe {
        if PMM_BITMAP.is_null() {
            return 0;
        }

        /* Scan from the hint forward, then wrap. */
        for pass in 0..2 {
            let (start, end) = if pass == 0 {
                (PMM_FIRST_FREE_HINT, PMM_TOTAL_PAGES)
            } else {
                (0, PMM_FIRST_FREE_HINT)
            };
            for p in start..end {
                if !bitmap_test(p) {
                    bitmap_set(p);
                    PMM_FREE_COUNT -= 1;
                    PMM_FIRST_FREE_HINT = p + 1;
                    return p * PAGE_SIZE;
                }
            }
        }
    }
    crate::kprintln!("[mm] PMM: out of physical memory");
    0
}

/// Marks the 4 KiB page at physical address `paddr` (must be
/// PAGE_SIZE-aligned) as free. Silently ignores 0 / misaligned.
pub fn free_page(paddr: u64) {
    unsafe {
        if paddr == 0 || paddr % PAGE_SIZE != 0 {
            return;
        }
        let p = paddr / PAGE_SIZE;
        if p >= PMM_TOTAL_PAGES || !bitmap_test(p) {
            return; /* out of range or already free */
        }
        bitmap_clear(p);
        PMM_FREE_COUNT += 1;
        if p < PMM_FIRST_FREE_HINT {
            PMM_FIRST_FREE_HINT = p;
        }
    }
}

/// Number of currently-free 4 KiB pages.
pub fn free_page_count() -> u64 {
    unsafe { PMM_FREE_COUNT }
}

/// Total number of pages tracked by the bitmap.
pub fn total_pages() -> u64 {
    unsafe { PMM_TOTAL_PAGES }
}
